docker/scripts/p4runtime_switch.py

```diff
@@ -0,0 +1,321 @@
+        flow['table'], match_str, flow['action_name'], params)
+
+# object hook for josn library, use str instead of unicode object
+# https://stackoverflow.com/questions/956867/how-to-get-string-objects-instead-of-unicode-from-json
+def json_load_byteified(file_handle):
+    return _byteify(json.load(file_handle, object_hook=_byteify),
+                    ignore_dicts=True)
+def _byteify(data, ignore_dicts=False):
+    # if this is a unicode string, return its string representation
+    if isinstance(data, unicode):
+        return data.encode('utf-8')
+    # if this is a list of values, return list of byteified values
+    if isinstance(data, list):
+        return [_byteify(item, ignore_dicts=True) for item in data]
+    # if this is a dictionary, return dictionary of byteified keys and values
+    # but only if we haven't already byteified it
+    if isinstance(data, dict) and not ignore_dicts:
+        return {
+            _byteify(key, ignore_dicts=True): _byteify(value, ignore_dicts=True)
+            for key, value in data.iteritems()
+        }
+    # if it's anything else, return it in its original form
+    return data
+
+class P4RuntimeSwitch(P4Switch):
+    "BMv2 switch with gRPC support"
+    next_grpc_port = 50051
+    next_thrift_port = 9090
+
+    def __init__(self, name, sw_path = None,
+                 grpc_port = None,
+                 thrift_port = None,
+                 pcap_dump = False,
+                 log_console = False,
+                 start_controller = True,
+                 program = None,
+                 verbose = False,
+                 device_id = None,
+                 enable_debugger = False,
+                 log_file = None,
+                 **kwargs):
+        Switch.__init__(self, name, **kwargs)
+        assert (sw_path)
+        self.sw_path = sw_path
+        # make sure that the provided sw_path is valid
+        pathCheck(sw_path)
+
+        self.program = program
+        json_path = None
+        p4info_path = None
+        if self.program:
+            json_path = self.program.json()
+            if self.program.p4info():
+                p4info_path = self.program.p4info()
+
+        if json_path is not None:
+            # make sure that the provided JSON file exists
+            if not os.path.isfile(json_path):
+                error("Invalid JSON file.\n")
+                exit(1)
+            self.json_path = json_path
+        else:
+            self.json_path = None
+
+        if p4info_path is not None:
+            if not os.path.isfile(p4info_path):
+                error("Invalid P4Info file.\n")
+                exit(1)
+            self.p4info_path = p4info_path
+        else:
+            self.p4info_path = None
+
+        if grpc_port is not None:
+            self.grpc_port = grpc_port
+        else:
+            self.grpc_port = P4RuntimeSwitch.next_grpc_port
+            P4RuntimeSwitch.next_grpc_port += 1
+
+        if thrift_port is not None:
```

Comment:
simple_switch_grpc does not support the Thrift RPC server by default

Reply:
I think whoever wrote the `P4RuntimeSwitch` class based it on `P4Switch`. As you mentioned, the `simple_switch_grpc` target doesn't support Thrift *by default*. However, according to the target's README, it could support Thrift, so I think we should keep this code in.